grpc-sys/build.rs

```diff
@@ -12,14 +12,170 @@
+    }
+}
+
+#[cfg(feature = "static-link")]
+mod imp {
+    use std::env;
+    use std::fs::DirBuilder;
+    use std::path::{Path, PathBuf};
+    use std::process::Command;
+
+    use super::GRPC_VERSION;
+
+    const ZLIB_VERSION: &'static str = "1.2.8";
+    const BORINGSSL_GIT_HASH: &'static str = "78684e5b222645828ca302e56b40b9daff2b2d27";
+    const CARES_VERSION: &'static str = "cares-1_12_0";
+
+    fn wget(url: &str, out: &str) -> Result<(), String> {
+        Command::new("wget")
+            .args(&["-q", "-c", "-O", out, url])
+            .status()
+            .map_err(|err| format!("wget execute failed: {}", err))
+            .and_then(|status| {
+                if status.success() {
+                    Ok(())
+                } else {
+                    Err(format!("wget exit with {}", status))
+                }
+            })
+    }
+
+    fn tar_xf(file: &str) -> Result<(), String> {
+        Command::new("tar")
+            .args(&["zxf", file])
+            .status()
+            .map_err(|err| format!("tar execute failed: {}", err))
+            .and_then(|status| {
+                if status.success() {
+                    Ok(())
+                } else {
+                    Err(format!("tar exit with {}", status))
+                }
+            })
+    }
+
+    pub fn build_or_link_grpc(cc: &mut ::gcc::Config) {
+        let cur_dir = env::current_dir().expect("Can't access current working directory");
+        let crate_dir = cur_dir.as_path().display();
+
+        let out_dir = env::var("OUT_DIR").expect("Can't access OUT_DIR");
+        for sub_dir in &["grpc", "zlib", "ssl", "cares"] {
+            DirBuilder::new().recursive(true).create(format!("{}/{}", out_dir, sub_dir)).unwrap();
+        }
+
+        if !Path::new(&format!("grpc-{}", GRPC_VERSION)).exists() {
+            wget(&format!("https://github.com/grpc/grpc/archive/v{}.tar.gz",
+                          GRPC_VERSION),
+                 "grpc.tar.gz")
+                .and_then(|_| tar_xf("grpc.tar.gz"))
+                .unwrap();
+        }
+
+        if !Path::new(&format!("zlib-{}", ZLIB_VERSION)).exists() {
+            wget(&format!("https://github.com/madler/zlib/archive/v{}.tar.gz",
+                          ZLIB_VERSION),
+                 "zlib.tar.gz")
+                .and_then(|_| tar_xf("zlib.tar.gz"))
+                .unwrap();
+        }
+
+        if !Path::new(&format!("boringssl-{}", BORINGSSL_GIT_HASH)).exists() {
+            wget(&format!("https://github.com/google/boringssl/archive/{}.tar.gz",
+                          BORINGSSL_GIT_HASH),
+                 "boringssl.tar.gz")
+                .and_then(|_| tar_xf("boringssl.tar.gz"))
+                .unwrap();
+        }
+
+        let cares_path = PathBuf::from(format!("grpc-{}/third_party/cares/cares", GRPC_VERSION));
+        if !cares_path.with_file_name("README.md").exists() {
+            wget(&format!("https://github.com/c-ares/c-ares/archive/{}.tar.gz",
```

Comment:
you can't handle every type of error. What's the outmost border?
Since `git-submodule` plan was rejected.
my plan was:

cargo build should works, if it not works, cargo clean; cargo build should works.

+TODO: should build in OUT_DIR